File: generators/app/templates/middle-tier/src/main/kotlin/vertx/service/UserService.kt
package <%= packageName %>.service

import io.vertx.core.AbstractVerticle
import io.vertx.core.eventbus.Message
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject

import <%= packageName %>.domain.User
import <%= packageName %>.repository.UserRepository

/**
 * Business service dealing with Users.
 */
class UserService(private val userRepository: UserRepository) : AbstractVerticle() {

    override fun start() {
        vertx.eventBus().consumer<Any>("users.all").handler { this.users(it) }
        vertx.eventBus().consumer<Any>("users.one").handler { this.user(it) }
    }

    private fun users(message: Message<Any>) {
        val users = userRepository.readUsers()

        // We manually construct an array of JSON objects from the list since Json.encode would do some escaping of
        // quote characters that we don't want - an alternative would be to create a wrapper object the for list (but at
        // least with streams this is not so bad)
        message.reply(JsonArray(users.map { JsonObject.mapFrom(it) }))
    }

    private fun user(message: Message<Any>) {
        val username = message.body().toString()
        val user: User? = userRepository.readUser(username)
        if (user != null) message.reply(JsonObject.mapFrom(user)) else message.fail(404, "No such user")
    }

}

File: generators/app/templates/middle-tier/src/main/scala/vertx/service/UserService.scala
package <%= packageName %>.service

import io.vertx.core.AbstractVerticle
import io.vertx.core.eventbus.Message
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject

import <%= packageName %>.domain.User
import <%= packageName %>.repository.UserRepository

/**
  * Business service dealing with Users.
  */
class UserService(val userRepository: UserRepository) extends AbstractVerticle {

    override def start(): Unit = {
        vertx.eventBus.consumer("users.all").handler(this.users)
        vertx.eventBus.consumer("users.one").handler(this.user)
    }

    private def users(message: Message[AnyRef]): Unit = {
        val json = userRepository.readUsers
            .map { user => toJson(user) }
            .foldLeft(new JsonArray) { (arr, user) => arr.add(user) }

        message.reply(json)
    }

    private def user(message: Message[AnyRef]): Unit = {
        userRepository.readUser(message.body.toString) match {
            case Some(user) => message.reply(toJson(user))
            case None => message.fail(404, "No such user")
        }
    }

    private def toJson(user: User): JsonObject = {
        new JsonObject()
            .put("username", user.username)
            .put("name", user.name)
    }

}
